sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StatePartitionWriter.scala

```diff
@@ -123,6 +159,12 @@ class StatePartitionAllColumnFamiliesWriter(
     val valueRow = new UnsafeRow(columnFamilyToValueSchemaLenMap(colFamilyName))
     valueRow.pointTo(valueBytes, valueBytes.length)
 
-    stateStore.put(keyRow, valueRow, colFamilyName)
+    if (columnFamilyToSchemaMap(colFamilyName).useMultipleValuesPerKey) {
+      // if a column family useMultipleValuesPerKey (e.g. ListType), we will
+      // write with 1 put followed by merge
+      stateStore.merge(keyRow, valueRow, colFamilyName)
```

Comment:
This is wrong right. You said it in your comment here, to do 1 put and then merge for the rest. But you are only doing merge for everything. You can use `stateStore.keyExists(keyRow, colFamilyName)` to determine whether to put or merge.

Reply:
@zifeif2 merge alone won't work in all cases. The expectation is Put first before merges